Develop

GanenkovNA opened the pull request.

GanenkovNA committed:
Добавил зависимости в pom

GanenkovNA committed:
Добавил описание задания

GanenkovNA committed:
Добавил Project Lombok

GanenkovNA committed:
Добавил Jackson Databind

GanenkovNA committed:
Создал сущность для курьера (`CourierEntity`) и DTO для API-запросов:
* `CourierCreationRequestDto`
* `CourierLoginRequestDto`
* `CourierLoginResponseDto`

GanenkovNA committed:
Создал базовый класс для тестирования запросов `CourierBase`, связанных с курьером.
Включает в себя следующие методы:
* `createValidCourierEntity()` - создание корректной сущности курьера;
* `addNewCourier()` - создание курьера;
* `loginCourier()` - логин курьера;
* `deleteCourier()` - удаление курьера;
* `*AndCheck()` - проверки выполнения операций выше.

GanenkovNA committed:
Добавил позитивную проверку создания курьера

GanenkovNA committed:
Добавил позитивную проверку логина курьера

GanenkovNA committed:
Добавил позитивную проверку удаления курьера

GanenkovNA committed:
Добавил используемые инструменты

GanenkovNA committed:
Добавил аннотации для Allure

GanenkovNA committed:
Добавил SLF4J Simple для логирования

GanenkovNA committed:
В коммите код нерабочий, но мне надо его сделать, ибо сижу уже больше 12 часов.
Что было сделано:
* Тесты
    * Были созданы отдельные методы для проверки статус кода (`assertStatusCode`) и значений в теле (`assertBody`)
    * Добавлены проверки на создание и логин курьера (чтобы не гонять в пустую)
* Допилено логирование
    * заменён `slf4j-simple` на `backlog` (его конфиг в `logback-test.xml`)
    * созданы DTO для запроса ('HttpRequestDto'), ответа (`HttpResponseDto`) и их отформатированного вывода (`HttpExchangeDto`)
    * Создан фильтр для перехвата запросов/ответов (`ExchangeCaptureFilter`)
* Allure:
    * сделан вывод запросов и ответов в Allure (`AllureLogger`)
    * добавлены аннотации
Наверное есть ещё что-то, но я так заебался...

GanenkovNA committed:
Рефактор кода

GanenkovNA committed:
Добавил отступы в лог)))0)0)

GanenkovNA committed:
Добавил позитивный тест на создание курьера с отправкой всех полей

GanenkovNA committed:
Добавил позитивные тесты:
* логин курьера с отправкой всех полей
* удаление существующего курьера

GanenkovNA committed:
Вынес логирование в `before`, `test` и `after` в класс `ScooterBase`

GanenkovNA committed:
1. Рефактор кода (много рефактора...)
2. Добавил негативные тесты на дубликат курьера

GanenkovNA committed:
Добавил настройку UTF-8 для кириллицы

GanenkovNA committed:
Изменил `isCreated` для возможности класть `null`

GanenkovNA committed:
Добавил метод для "аварийного" удаления курьера

GanenkovNA committed:
1. Вынес константы в переменные класса
2. Рефактор (вызов "аварийного" удаления)

GanenkovNA committed:
1. Исправил метод `tryToDeleteInvalidCourierInCaseOfTestFail`
2. Обновил аннотации и логи

GanenkovNA committed:
Создал тесты для проверки невозможности создания курьера с отсутствующим полем

GanenkovNA committed:
Создал тесты для проверки невозможности создания курьера с `null` полем

GanenkovNA committed:
Перенёс негативные тесты на создание курьера

GanenkovNA committed:
Создал проверки невозможности логина курьера с неверным паролем

GanenkovNA committed:
Добавлена константа `DEFAULT_LENGTH_OF_GENERATED_STRING`

GanenkovNA committed:
Добавил проверка невозможности логина несуществующего курьера

GanenkovNA committed:
Добавил проверку невозможности логина курьера с отсутствующим полем

GanenkovNA committed:
Создал `ApiClient`

GanenkovNA committed:
Обновил `CourierService` на использование `ApiClient`

GanenkovNA committed:
Обновил `createValidCourierEntity` на создание рандомного пользователя

GanenkovNA committed:
1. Удалил ненужный импорт
2. Исправил `createValidCourierEntity`

GanenkovNA committed:
Добавил в `ApiClient` тайм-ауты и их обработку и логирование

GanenkovNA committed:
Переработал `LoginCourierWithMissingFieldTest`

GanenkovNA committed:
Добавил проверку невозможности логина курьера с null полем

GanenkovNA committed:
Перенёс `DEFAULT_LENGTH_OF_GENERATED_STRING` в `ScooterBase`

GanenkovNA committed:
Добавил метод для PUT

GanenkovNA committed:
Добавил класс для генерации даты

GanenkovNA committed:
Добавил сущность и DTO для заказа

GanenkovNA committed:
Добавил базовый и обслуживающий классы для заказа

GanenkovNA committed:
коммит для переноса на комп

GanenkovNA committed:
Добавил DTO и методы для отправки запроса на получение данных по заказу по `track` и проверки этого запроса

GanenkovNA committed:
Добавил в тест заказ с одновременным указанием цветов и добавил в очистку после теста проверку нахождения заказа в бд

GanenkovNA committed:
Добавил тест с пустым значением поля `color`

GanenkovNA committed:
Добавил аннотации для Allure

GanenkovNA committed:
Добавил тест на создание заказа с отсутствующим полем `color`

GanenkovNA committed:
Мелкие исправления

GanenkovNA committed:
Мелкие исправления

GanenkovNA committed:
Добавил тест на получение списка

GanenkovNA committed:
add allure report

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/CourierBase.java:
```diff
@@ -0,0 +1,137 @@
+        logger.debug("Отправлен запрос на создание курьера {}", courier.getLogin());
+
+        assertStatusCode(response,
+                201,
```
⚠️Можно улучшить. Вместо числовых кодов ответов, лучше использовать коды из библиотеки static org.apache.http.HttpStatus, так как они там строковые и несут смысл, что повышает читаемость и не надо вспоминать, что есть что. Пример кода, statusCode(SC_CREATED).

GanenkovNA replied:
Исправлено

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/create/CreateCourierWithMissingFieldTest.java:
```diff
@@ -0,0 +1,116 @@
+        });
+    }
+
+    @DisplayName("Проверка кода ответа (`" + EXPECTED_STATUS_CODE + "`)")
```
⚠️Можно улучшить. В allure есть еще одна аннотация для описания теста Description. Добавь ее, пожалуйста, во всех тестах проекта

GanenkovNA replied:
Исправлено

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/create/CreateCourierWithMissingFieldTest.java:
```diff
@@ -0,0 +1,116 @@
+
+    @DisplayName("Проверка кода ответа (`" + EXPECTED_STATUS_CODE + "`)")
+    @Test
+    public void shouldNotCreateCourierWithMissingFieldAndVerifyStatusCode() {
```
⚠️Можно улучшить. В названии метода необходимо использовать слово Test. Проверь, соблюдается ли данное правило для всех созданных тестов.

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/create/CreateCourierWithMissingFieldTest.java:
```diff
@@ -0,0 +1,116 @@
+
+    @DisplayName("Проверка параметра `message` в ответе")
+    @Test
+    public void shouldNotCreateCourierWithMissingFieldAndVerifyMessage(){
```
⛔️Нужно исправить.  Не нужен отдельный текст на проверку кода. Код и тело ответа проверяем в одном тест-кейсе

GanenkovNA replied:
Исправлено

okryukova commented:
⛔️Нужно исправить.  Часть тестов упала при запуске. Проверь , пожалуйста

![image](https://github.com/user-attachments/assets/6c4fa156-7625-49ca-8f43-b96d22709d6b)
![image](https://github.com/user-attachments/assets/2066aa3e-a905-4ac1-8700-56e204c958b2)
![image](https://github.com/user-attachments/assets/f6d84f79-98bc-4682-9044-926dadbeaf5d)

GanenkovNA committed:
Разбил на папки

GanenkovNA committed:
`CreateValidCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`

GanenkovNA committed:
`DeleteValidCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `verifyResponseWithLog` за ненадобностью

GanenkovNA committed:
`LoginValidCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `verifyResponseWithLog` за ненадобностью

GanenkovNA committed:
Revert "`LoginValidCourierTest`"

This reverts commit 1336a71bd461969645f65079f10c5c0f063d15d2.

GanenkovNA committed:
`LoginValidCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `verifyResponseWithLog` за ненадобностью

GanenkovNA committed:
`LoginValidCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `verifyResponseWithLog` за ненадобностью

GanenkovNA committed:
`CreateCourierWithMissingFieldTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `createCourierWithMissingFieldAndVerify` за ненадобностью

GanenkovNA committed:
`CreateCourierWithNullFieldTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `createCourierWithNullFieldAndVerify` за ненадобностью

GanenkovNA committed:
`CreateDuplicateOfCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значения статус кодов на константы из `org.apache.http`
4. Удалил методы `verifyCreateResponseWithLog` и `verifyLoginResponseWithLog` за ненадобностью

GanenkovNA committed:
`LoginCourierWithInvalidPasswordTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `loginCourierWithInvalidPasswordAndVerify` за ненадобностью

GanenkovNA committed:
`LoginCourierWithMissingFieldTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `loginCourierWithMissingFieldAndVerify` за ненадобностью

GanenkovNA committed:
`LoginCourierWithNullFieldTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `loginCourierWithNullFieldAndVerify` за ненадобностью

GanenkovNA committed:
`LoginCourierWithNullFieldTest`
Удалил методы `verifyResponse` и `deleteDuplicateOfCourier` за ненадобностью

GanenkovNA committed:
`LoginNonExistentCourierTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `loginNonExistentCourierAndVerify` за ненадобностью

GanenkovNA committed:
Заменил значение статус кодов на константы из `org.apache.http`

GanenkovNA committed:
Удалил `loginAnyCourierAndVerify`

GanenkovNA committed:
Заменил значение статус кодов на константы из `org.apache.http`

GanenkovNA committed:
`CreateValidOrderWithAnyColorTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `createValidOrderWithOnlyColorAndVerify` за ненадобностью

GanenkovNA committed:
`CreateValidOrderWithMissingColorFieldTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `createValidOrderWithMissingColorFieldAndVerify` за ненадобностью

GanenkovNA committed:
`GetListOfOrdersTest`
1. Объединил тесты на проверку статус кода и тела
2. Добавил `@Description`
3. Заменил значение статус кода на константу из `org.apache.http`
4. Удалил метод `returnListOfOrdersAndVerify` за ненадобностью

GanenkovNA committed:
Обновил отчёт Allure

GanenkovNA committed:
1. Тест на создание курьера с пропущенным полем `firstName` переделан в позитивный (`CreateValidCourierWithMissingFirstNameFieldTest`)
2. Метод `addNewCourierWithMissingFields` перенесён в `CourierBase`
3. Создана отдельная директория для позитивных тестов на создание курьера

GanenkovNA committed:
Изменил выбрасываемое исключение в методе `send` класса `ApiClient` для того, чтобы тесты отображались как `failed`, а не `broken`

GanenkovNA committed:
Добавил отчёт Allure

okryukova commented on src/test/java/ru/yandex/praktikum/infrastructure/RandomStringGenerator.java:
```diff
@@ -0,0 +1,17 @@
+
+public class RandomStringGenerator {
+    public static String generateRandomString(int length) {
+        String characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
```
⚠️Можно улучшить. Генератор данных это отлично, но лучше использовать готовые решения JavaFaker/DataFaker, так как это сокращает нашу работу.

GanenkovNA replied:
Понял, спасибо за совет! В дипломном проекте обязательно воспользуюсь!

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/create/CreateCourierWithNullFieldTest.java:
```diff
@@ -0,0 +1,87 @@
+
+@DisplayName("Проверка невозможности создания курьера с `null` полем")
+@RunWith(Parameterized.class)
+public class CreateCourierWithNullFieldTest extends CourierBase {
```
⛔️Нужно исправить.  Данный класс дублирует тесты из класса CreateCourierWithMissingFieldTest
ТД одинаковы. Если я не права, подскажи в чем разница
![image](https://github.com/user-attachments/assets/dbc07b7a-78b9-4ace-84a0-d4536cbfabbf)

GanenkovNA replied:
Исправлено

GanenkovNA replied:
При отправке запроса c полем с значением `null` мы получаем такое тело:
```json
{
    "login": null,
    "password": "password"
}
```

При отправке запроса без поля мы получаем такое тело:
```json
{
    "password": "password"
}
```

Это разные тесты, в которых поведение бэка может различаться (хотя и не должно). Как пример - злополучные `LoginCourierWithMissingFieldTest` и `LoginCourierWithNullFieldTest`

GanenkovNA replied:
Объяснение, если вопрос связан с кодом.
В случае с `null` полем - я просто сериализую объект, у которого параметр имеет такое значение, и отправляю стандартным методом.
В случае с пропущенным полем - я создаю строку без этого поля c помощью `ObjectMapper` и отправляю другим методом (оканчивается на `WithMissingField`).

okryukova replied:
Никита, спасибо большое за пояснения. У тебя отличная работа. Используешь подходы и методы не описанные в курсе. Глубоко разобрался в задаче. Молодец!

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/create/CreateCourierWithNullFieldTest.java:
```diff
@@ -0,0 +1,87 @@
+        return new Object[][] {
+                {null, generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING), generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING)},
+                {generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING), null, generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING)},
+                {generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING), generateRandomString(DEFAULT_LENGTH_OF_GENERATED_STRING), null}
```
⛔️Нужно исправить.  firstName необязательное поле. Создать курьера можно и без него

GanenkovNA replied:
Исправлено

okryukova commented on src/test/java/ru/yandex/praktikum/scooter_test/courier_test/negative/login/LoginCourierWithMissingFieldTest.java:
```diff
@@ -0,0 +1,108 @@
+                {"1", null}
+        };
+    }
+
```
![image](https://github.com/user-attachments/assets/db800814-8a7c-4e2c-88ed-472a07251ee7)


⛔️Нужно исправить.  Тест  на логин без пароля падает с ошибкой  java.net.SocketTimeoutException: Read timed out. У тебя не отправляется запрос. Тебе нужно проверить метод отправки запроса без обязательного поля, возможно там есть ошибка

GanenkovNA replied:
Замечание снято

okryukova replied:
Студенты редко находят этот дефект. Ты отлично поработал!

GanenkovNA committed:
Удалил тесты на поле `firstName` (проморгал)

GanenkovNA committed:
Убрал из запроса `id`...

GanenkovNA committed:
Перенёс методы, отвечающие просто за запросы, из `CourierBase` в `CourierService`

GanenkovNA committed:
Обновил отчёт Allure
